library/Freckle/App/Stats.hs

```diff
@@ -111,14 +115,33 @@ withStatsClient StatsSettings {..} f = do
   if amsEnabled
     then do
       tags <- (amsTags <>) <$> getEcsMetadataTags
-
+      gauges <- liftIO $ newIORef mempty
       Datadog.withDogStatsD amsSettings $ \client ->
         -- Add the tags to the thread context so they're present in all logs
         withThreadContext (map toPair tags)
-          $ f StatsClient { scClient = client, scTags = tags }
-    else f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags }
+          $ f StatsClient { scClient = client, scTags = tags, scGauges = gauges }
+    else do
+      gauges <- liftIO $ newIORef mempty
+      f $ StatsClient { scClient = Datadog.Dummy, scTags = amsTags, scGauges = gauges }
   where toPair = bimap (fromString . unpack) String
 
+withGauge :: (MonadReader app m, HasStatsClient app, MonadUnliftIO m) => Text -> m a -> m a
+withGauge name f = do
+  gaugesRef <- asks $ scGauges . view statsClientL
+  gaugesMap <- liftIO $ readIORef gaugesRef
+  gaugeValue <- liftIO $ maybe (createNewEKG gaugesRef gaugesMap) pure (HashMap.lookup name gaugesMap)
```

Comment:
Uh-oh. I may have led you astray.

This is super duper racy. If multiple threads interleave `readIORef` and `writeIORef`, we would under-increment. Given how hot `runDB` is, that seems like a realistic problem.

Unfortunately, I'm not sure how to solve it. If we were OK with an unnecessary `EKG.new`, you could put the lookup-or-insert inside an `atomicModifyIORef'` to address the race condition, but now you'd be forcibly serializing all `runDB` calls to lock on the Gauge. Again, given how hot `runDB` is, I don't know that that is a good idea to even "ship it and see".

So, I just don't think we can dynamically create gauges. I still think we can keep the single `HasStatsClient` interface with an `scGauges` member. But it would have to be a statically defined structure, either a non-`IORef` `HashMap` or we go ahead with the record type idea presented elsewhere.

I'd be OK with this, since we only use it internally in `freckle-app` now for `runDB`. Giving up the feature of end-users dynamically wrapping things in gauges seems acceptable since we don't even have a use-case for that ourselves.

WDYT?

Reply:
All the racing would happen at start up, but even then I think we'd get issues. I think just a static `scGauge :: Gauge` would be the simplest solution, especially given the lack of use case for more types of gauges.

Reply:
If you wanted a more complicated interface that allowed apps to hold their gauges abstractly, we could convert it to `scGauge: a` and require `class NamedGauge a where name :: Text` on our `class NamedGauge a => HasStatsClient a`.

Reply:
> All the racing would happen at start up

I suppose that's better, but I still think pretty problematic. We boot up a process as part of scaling, it's grabs some N requests at the same time in a thread each, then a ton of them call `runDB` basically immediately. Will that set the gauge up as wrong forever? Will we see negative counts?

> I think just a static `scGauge :: Gauge` would be the simplest solution

Sounds good. Minor nit, but I'd still do `scGauges` (plural), even if the db-connection gauge is the only one we have now.

Adding more (static) gauges later feels likely, _dynamically creating_ gauges is the feature I'm OK to give up.

Reply:
Yes, it probably would mess up the count forever, you could quite easily lose a bunch of adds and end up negative.

Then you probably want `scGauges :: Gauges` where `data Gauges = { dbConnections :: EKG.Gauge }` so any refactoring down the line doesn't require extensive changes.

Reply:
Yeah, that works. I think I called it `StatsClientGauges { scgDbConnections }` originally, just to go with the general theme of the module, but un-prefixed may even be better considering things are generally qualified anyway,

```hs
Stats.withGauge Stats.dbConnections $ do
  ...
```